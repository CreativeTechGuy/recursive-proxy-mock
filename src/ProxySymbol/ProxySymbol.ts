/* eslint-disable @typescript-eslint/naming-convention */
/**
 * Object containing a Symbol for each of the Proxy handler functions. Used to construct a ProxyPath.
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy#handler_functions} for a list of Proxy handler names
 */
export const ProxySymbol = {
    APPLY: Symbol("apply"),
    CONSTRUCT: Symbol("construct"),
    DEFINE_PROPERTY: Symbol("defineProperty"),
    DELETE_PROPERTY: Symbol("deleteProperty"),
    GET_OWN_PROPERTY_DESCRIPTOR: Symbol("getOwnPropertyDescriptor"),
    GET_PROTOTYPE_OF: Symbol("getPrototypeOf"),
    HAS: Symbol("has"),
    IS_EXTENSIBLE: Symbol("isExtensible"),
    OWN_KEYS: Symbol("ownKeys"),
    PREVENT_EXTENSIONS: Symbol("preventExtensions"),
    SET: Symbol("set"),
    SET_PROTOTYPE_OF: Symbol("setPrototypeOf"),
} as const;

export const ProxyStackSymbol = Symbol("proxyStack");
