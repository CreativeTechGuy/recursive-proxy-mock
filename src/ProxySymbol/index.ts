export { ProxySymbol, ProxyStackSymbol } from "./ProxySymbol";
