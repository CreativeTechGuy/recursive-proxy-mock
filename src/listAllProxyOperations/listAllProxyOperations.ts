import { isRecursiveProxyMock } from "~/isRecursiveProxyMock";
import type { ProxyData } from "~/proxyTypes";
import { developmentLog } from "~/utils/developmentLog";
import { getProxyStack } from "~/utils/getProxyStack";

/**
 * @param proxy - the root proxy object returned from {@link recursiveProxyMock}
 * @returns Array of {@link ProxyData} with one entry for every interaction on the proxy object
 */
export function listAllProxyOperations(proxy: unknown): ProxyData[] {
    if (!isRecursiveProxyMock(proxy)) {
        developmentLog("Must pass an object created with `recursiveProxyMock()`. Instead received:", proxy);
        return [];
    }
    return getProxyStack(proxy);
}
