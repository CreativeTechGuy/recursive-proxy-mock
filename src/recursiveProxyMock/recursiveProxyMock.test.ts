/* eslint-disable no-new */
/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { JSDOM } from "jsdom";
import type { hasPathBeenVisited as HasPathBeenVisitedType } from "~/hasPathBeenVisited";
import { listAllProxyOperations } from "~/listAllProxyOperations";
import { ProxySymbol } from "~/ProxySymbol";
import type { recursiveProxyMock as RecursiveProxyMockType } from "~/recursiveProxyMock";

describe("recursiveProxyMock", () => {
    let recursiveProxyMock: typeof RecursiveProxyMockType;
    let hasPathBeenVisited: typeof HasPathBeenVisitedType;
    beforeEach(async () => {
        jest.resetModules();
        recursiveProxyMock = (await import("~/recursiveProxyMock")).recursiveProxyMock;
        hasPathBeenVisited = (await import("~/hasPathBeenVisited")).hasPathBeenVisited;
    });

    test("example: do literally anything", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            mock.a.b.c;
            mock.d().test();
            mock.person.name = "Jason";
            new mock.MyClass();
            const value = mock.getThing("yes", 42);
            value.something.else = true;
            value.something.else().and.now().val = "WOW";
        }).not.toThrow();
    });

    test("example: HTML5 Canvas mock", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            const {
                window: { document },
                window,
            } = new JSDOM();
            window.HTMLCanvasElement.prototype.getContext = mock;

            const canvas = document.createElement("canvas");
            const context = canvas.getContext("webgl")!;
            context.clear(context.COLOR_BUFFER_BIT);
        }).not.toThrow();
        expect(hasPathBeenVisited(mock, [ProxySymbol.APPLY])).toStrictEqual(true);
        expect(hasPathBeenVisited(mock, [ProxySymbol.APPLY, "clear", ProxySymbol.APPLY])).toStrictEqual(true);
    });

    test("example: jQuery entire library mock", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            const $ = mock;
            $("div").append("<p>Content</p>").css("color", "blue").click();
        }).not.toThrow();
        expect(hasPathBeenVisited(mock, [ProxySymbol.WILDCARD, "click", ProxySymbol.APPLY])).toStrictEqual(true);
    });

    test("example: complex object mock", () => {
        const resMock = recursiveProxyMock();
        const reqMock = recursiveProxyMock([
            {
                path: ["session", "destroy", ProxySymbol.APPLY],
                value: (callback: () => void): void => {
                    callback();
                },
            },
        ]);
        expect(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            function logoutHandler(req: any, res: any): void {
                req.session.destroy(() => {
                    res.redirect("/");
                });
            }
            logoutHandler(reqMock, resMock);
        }).not.toThrow();
        expect(hasPathBeenVisited(resMock, ["redirect", ProxySymbol.APPLY])).toStrictEqual(true);
    });

    test("supports math operations on primitive", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            mock.test *= 10;
            // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
            mock.a.b = mock.c.d.e + (2 * mock.a.b) / 2;
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("supports spread as an array", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            mock.spread = [...mock.should.spread];
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: apply", () => {
        const fn = jest.fn(() => 7);
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.APPLY],
                value: (): number => {
                    return fn();
                },
            },
        ]);
        expect(mock.a.b() * 2).toStrictEqual(14);
        expect(fn).toHaveBeenCalled();
        expect(() => {
            mock.c().d();
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: construct", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.CONSTRUCT],
                value: class OverrideClass {
                    private readonly multiplyBy: number;
                    public constructor(multiplyBy: number) {
                        this.multiplyBy = multiplyBy;
                    }
                    public getValue(): number {
                        return 7 * this.multiplyBy;
                    }
                },
            },
        ]);
        expect(new mock.a.b(2).getValue()).toStrictEqual(14);
        expect(() => {
            new new mock.c().d();
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: defineProperty", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.DEFINE_PROPERTY],
                value: false,
            },
        ]);
        expect(() => {
            Object.defineProperty(mock.a, "b", {
                value: 7,
            });
        }).toThrow(TypeError);
        expect(() => {
            Object.defineProperty(mock.a, "c", {
                value: "potato",
            });
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: deleteProperty", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.DELETE_PROPERTY],
                value: false,
            },
        ]);
        expect(() => {
            delete mock.a.b;
        }).toThrow(TypeError);
        expect(() => {
            delete mock.c.d;
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: get", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b"],
                value: "potato",
            },
            {
                path: ["num", "val", Symbol.toPrimitive],
                value: 10,
            },
        ]);
        expect(mock.num.val * 2).toStrictEqual(20);
        expect(mock.a.b).toStrictEqual("potato");
        expect(mock.b.a).not.toBeInstanceOf(String);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: getOwnPropertyDescriptor", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.GET_OWN_PROPERTY_DESCRIPTOR],
                value: {
                    configurable: true,
                    enumerable: false,
                    writable: true,
                    value: 7,
                },
            },
        ]);
        expect(Object.getOwnPropertyDescriptor(mock.a, "b")).toStrictEqual({
            configurable: true,
            enumerable: false,
            writable: true,
            value: 7,
        });
        expect(() => {
            Object.getOwnPropertyDescriptor(mock.a, "c")!.value.d.e;
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: getPrototypeOf", () => {
        const mock = recursiveProxyMock();
        expect(Object.getPrototypeOf(mock.b.c)).toBeNull();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: has", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.HAS],
                value: false,
            },
        ]);
        expect("b" in mock.a).toStrictEqual(false);
        expect("c" in mock.a).toStrictEqual(true);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: isExtensible", () => {
        const mock = recursiveProxyMock();
        expect(Object.isExtensible(mock.a.b)).toStrictEqual(true);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: ownKeys", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.OWN_KEYS],
                value: ["po", "ta", "to"],
            },
            {
                path: ["a", "b", ProxySymbol.GET_PROTOTYPE_OF],
                value: null,
            },
        ]);
        expect(Object.keys(mock.a.b)).toStrictEqual(["po", "ta", "to"]);
        expect(Object.keys(mock.b.c)).toStrictEqual([]);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: preventExtensions", () => {
        const mock = recursiveProxyMock();
        expect(() => {
            Object.preventExtensions(mock.a.b);
        }).toThrow(TypeError);
        expect(() => {
            Object.preventExtensions(mock.b.c);
        }).toThrow(TypeError);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: set", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.SET],
                value: false,
            },
        ]);
        expect(() => {
            mock.a.b = 7;
        }).toThrow(TypeError);
        expect(() => {
            mock.b.c = "potato";
        }).not.toThrow();
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });

    test("proxy trap: setPrototypeOf", () => {
        const mock = recursiveProxyMock([
            {
                path: ["a", "b", ProxySymbol.SET_PROTOTYPE_OF],
                value: false,
            },
        ]);
        expect(() => {
            Object.setPrototypeOf(mock.a.b, {
                protoValue: 7,
            });
        }).toThrow(TypeError);
        expect(() => {
            Object.setPrototypeOf(mock.b.c, {
                protoValue: "potato",
            });
        }).not.toThrow(TypeError);
        expect(listAllProxyOperations(mock)).toMatchSnapshot();
    });
});
