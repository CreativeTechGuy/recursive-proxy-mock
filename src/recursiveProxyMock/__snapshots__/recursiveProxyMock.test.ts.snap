// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursiveProxyMock proxy trap: apply 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "args": [],
    "name": "apply",
    "parent": 2,
    "pathKey": [
      Symbol(apply),
    ],
    "self": 3,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "args": [],
    "name": "apply",
    "parent": 4,
    "pathKey": [
      Symbol(apply),
    ],
    "self": 5,
  },
  {
    "name": "get",
    "parent": 5,
    "pathKey": [
      "d",
    ],
    "prop": "d",
    "self": 6,
  },
  {
    "args": [],
    "name": "apply",
    "parent": 6,
    "pathKey": [
      Symbol(apply),
    ],
    "self": 7,
  },
]
`;

exports[`recursiveProxyMock proxy trap: construct 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "args": [
      2,
    ],
    "name": "construct",
    "parent": 2,
    "pathKey": [
      Symbol(construct),
    ],
    "self": 3,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "args": [],
    "name": "construct",
    "parent": 4,
    "pathKey": [
      Symbol(construct),
    ],
    "self": 5,
  },
  {
    "name": "get",
    "parent": 5,
    "pathKey": [
      "d",
    ],
    "prop": "d",
    "self": 6,
  },
  {
    "args": [],
    "name": "construct",
    "parent": 6,
    "pathKey": [
      Symbol(construct),
    ],
    "self": 7,
  },
]
`;

exports[`recursiveProxyMock proxy trap: defineProperty 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "descriptor": {
      "value": 7,
    },
    "name": "defineProperty",
    "parent": 1,
    "pathKey": [
      "b",
      Symbol(defineProperty),
    ],
    "prop": "b",
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 2,
  },
  {
    "descriptor": {
      "value": "potato",
    },
    "name": "defineProperty",
    "parent": 2,
    "pathKey": [
      "c",
      Symbol(defineProperty),
    ],
    "prop": "c",
  },
]
`;

exports[`recursiveProxyMock proxy trap: deleteProperty 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "deleteProperty",
    "parent": 1,
    "pathKey": [
      "b",
      Symbol(deleteProperty),
    ],
    "prop": "b",
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 2,
  },
  {
    "name": "deleteProperty",
    "parent": 2,
    "pathKey": [
      "d",
      Symbol(deleteProperty),
    ],
    "prop": "d",
  },
]
`;

exports[`recursiveProxyMock proxy trap: get 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "num",
    ],
    "prop": "num",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "val",
    ],
    "prop": "val",
    "self": 2,
  },
  {
    "name": "get",
    "parent": 2,
    "pathKey": [
      Symbol(Symbol.toPrimitive),
    ],
    "prop": Symbol(Symbol.toPrimitive),
    "self": 3,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 4,
  },
  {
    "name": "get",
    "parent": 4,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 5,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 6,
  },
  {
    "name": "get",
    "parent": 6,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 7,
  },
  {
    "name": "getPrototypeOf",
    "parent": 7,
    "pathKey": [
      Symbol(getPrototypeOf),
    ],
    "self": 8,
  },
]
`;

exports[`recursiveProxyMock proxy trap: getOwnPropertyDescriptor 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 1,
    "pathKey": [
      "b",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 3,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 3,
    "pathKey": [
      "c",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "name": "get",
    "parent": 4,
    "pathKey": [
      "d",
    ],
    "prop": "d",
    "self": 5,
  },
  {
    "name": "get",
    "parent": 5,
    "pathKey": [
      "e",
    ],
    "prop": "e",
    "self": 6,
  },
]
`;

exports[`recursiveProxyMock proxy trap: getPrototypeOf 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 2,
  },
  {
    "name": "getPrototypeOf",
    "parent": 2,
    "pathKey": [
      Symbol(getPrototypeOf),
    ],
    "self": 3,
  },
]
`;

exports[`recursiveProxyMock proxy trap: has 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "has",
    "parent": 1,
    "pathKey": [
      "b",
      Symbol(has),
    ],
    "prop": "b",
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 2,
  },
  {
    "name": "has",
    "parent": 2,
    "pathKey": [
      "c",
      Symbol(has),
    ],
    "prop": "c",
  },
]
`;

exports[`recursiveProxyMock proxy trap: isExtensible 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "isExtensible",
    "parent": 2,
    "pathKey": [
      Symbol(isExtensible),
    ],
  },
]
`;

exports[`recursiveProxyMock proxy trap: ownKeys 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "ownKeys",
    "parent": 2,
    "pathKey": [
      Symbol(ownKeys),
    ],
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "po",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "po",
    "self": 3,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "ta",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "ta",
    "self": 4,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "to",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "to",
    "self": 5,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "arguments",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "arguments",
    "self": 6,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "caller",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "caller",
    "self": 7,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 2,
    "pathKey": [
      "prototype",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "prototype",
    "self": 8,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 9,
  },
  {
    "name": "get",
    "parent": 9,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 10,
  },
  {
    "name": "ownKeys",
    "parent": 10,
    "pathKey": [
      Symbol(ownKeys),
    ],
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 10,
    "pathKey": [
      "arguments",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "arguments",
    "self": 11,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 10,
    "pathKey": [
      "caller",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "caller",
    "self": 12,
  },
  {
    "name": "getOwnPropertyDescriptor",
    "parent": 10,
    "pathKey": [
      "prototype",
      Symbol(getOwnPropertyDescriptor),
    ],
    "prop": "prototype",
    "self": 13,
  },
]
`;

exports[`recursiveProxyMock proxy trap: preventExtensions 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "preventExtensions",
    "parent": 2,
    "pathKey": [
      Symbol(preventExtensions),
    ],
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 3,
  },
  {
    "name": "get",
    "parent": 3,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "name": "preventExtensions",
    "parent": 4,
    "pathKey": [
      Symbol(preventExtensions),
    ],
  },
]
`;

exports[`recursiveProxyMock proxy trap: set 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "set",
    "parent": 1,
    "pathKey": [
      "b",
      Symbol(set),
    ],
    "prop": "b",
    "value": 7,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "set",
    "parent": 2,
    "pathKey": [
      "c",
      Symbol(set),
    ],
    "prop": "c",
    "value": "potato",
  },
]
`;

exports[`recursiveProxyMock proxy trap: setPrototypeOf 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 2,
  },
  {
    "name": "setPrototypeOf",
    "parent": 2,
    "pathKey": [
      Symbol(setPrototypeOf),
    ],
    "prototype": {
      "protoValue": 7,
    },
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 3,
  },
  {
    "name": "get",
    "parent": 3,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "name": "setPrototypeOf",
    "parent": 4,
    "pathKey": [
      Symbol(setPrototypeOf),
    ],
    "prototype": {
      "protoValue": "potato",
    },
  },
]
`;

exports[`recursiveProxyMock supports math operations on primitive 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "test",
    ],
    "prop": "test",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      Symbol(Symbol.toPrimitive),
    ],
    "prop": Symbol(Symbol.toPrimitive),
    "self": 2,
  },
  {
    "name": "set",
    "parent": 0,
    "pathKey": [
      "test",
      Symbol(set),
    ],
    "prop": "test",
    "value": 0,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 3,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "c",
    ],
    "prop": "c",
    "self": 4,
  },
  {
    "name": "get",
    "parent": 4,
    "pathKey": [
      "d",
    ],
    "prop": "d",
    "self": 5,
  },
  {
    "name": "get",
    "parent": 5,
    "pathKey": [
      "e",
    ],
    "prop": "e",
    "self": 6,
  },
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "a",
    ],
    "prop": "a",
    "self": 7,
  },
  {
    "name": "get",
    "parent": 7,
    "pathKey": [
      "b",
    ],
    "prop": "b",
    "self": 8,
  },
  {
    "name": "get",
    "parent": 8,
    "pathKey": [
      Symbol(Symbol.toPrimitive),
    ],
    "prop": Symbol(Symbol.toPrimitive),
    "self": 9,
  },
  {
    "name": "get",
    "parent": 6,
    "pathKey": [
      Symbol(Symbol.toPrimitive),
    ],
    "prop": Symbol(Symbol.toPrimitive),
    "self": 10,
  },
  {
    "name": "set",
    "parent": 3,
    "pathKey": [
      "b",
      Symbol(set),
    ],
    "prop": "b",
    "value": "0",
  },
]
`;

exports[`recursiveProxyMock supports spread as an array 1`] = `
[
  {
    "name": "get",
    "parent": 0,
    "pathKey": [
      "should",
    ],
    "prop": "should",
    "self": 1,
  },
  {
    "name": "get",
    "parent": 1,
    "pathKey": [
      "spread",
    ],
    "prop": "spread",
    "self": 2,
  },
  {
    "name": "get",
    "parent": 2,
    "pathKey": [
      Symbol(Symbol.iterator),
    ],
    "prop": Symbol(Symbol.iterator),
    "self": 3,
  },
  {
    "args": [],
    "name": "apply",
    "parent": 3,
    "pathKey": [
      Symbol(apply),
    ],
    "self": 4,
  },
  {
    "name": "get",
    "parent": 4,
    "pathKey": [
      "next",
    ],
    "prop": "next",
    "self": 5,
  },
  {
    "args": [],
    "name": "apply",
    "parent": 5,
    "pathKey": [
      Symbol(apply),
    ],
    "self": 6,
  },
  {
    "name": "get",
    "parent": 6,
    "pathKey": [
      "done",
    ],
    "prop": "done",
    "self": 7,
  },
  {
    "name": "set",
    "parent": 0,
    "pathKey": [
      "spread",
      Symbol(set),
    ],
    "prop": "spread",
    "value": [],
  },
]
`;
